pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -1750,3 +1769,85 @@ func (r *ReconcileClusterDeployment) deleteStaleProvisions(provs []*hivev1.Clust
 		}
 	}
 }
+
+// getAllSyncSetInstances return all syncset instances for a cluster deployment
+func (r *ReconcileClusterDeployment) getAllSyncSetInstances(cd *hivev1.ClusterDeployment) ([]*hivev1.SyncSetInstance, error) {
+
+	// get all syncset instances for the clusterdeployment namespace
+	ssiList := &hivev1.SyncSetInstanceList{}
+	err := r.Client.List(context.TODO(), ssiList, client.InNamespace(cd.Namespace))
+	if err != nil {
+		return nil, err
+	}
+
+	syncSetInstances := []*hivev1.SyncSetInstance{}
+	for i, syncSetInstance := range ssiList.Items {
+		if syncSetInstance.Spec.ClusterDeployment.Name == cd.Name {
+			syncSetInstances = append(syncSetInstances, &ssiList.Items[i])
+		}
+	}
+	return syncSetInstances, nil
+}
+
+// checkFailureSyncCondition return true when syncsetinstance status condition has hivev1.ApplyFailureSyncCondition
+func checkFailureSyncCondition(syncSetInstances []*hivev1.SyncSetInstance) bool {
+	for _, syncSetInstance := range syncSetInstances {
+		for _, condition := range syncSetInstance.Status.Conditions {
+			if condition.Type == hivev1.ApplyFailureSyncCondition {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// setSyncSetFailedCondition returns true when it sets the hivev1.SyncSetFailedCondition
+// hivev1.SyncSetFailedCondition is set if any syncsetinstance has status condition type as hivev1.ApplyFailureSyncCondition will return true
+func (r *ReconcileClusterDeployment) setSyncSetFailedCondition(cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) (bool, error) {
+
+	cdLog.Info("Checking syncset instance status") //Temp
+	// get all syncset instances for this cluster deployment
+	syncSetInstances, err := r.getAllSyncSetInstances(cd)
+	if err != nil {
+		cdLog.WithError(err).Error("Unable to list related syncset instances for cluster deployment")
+		return false, err
+	}
+
+	isfailedCondition := checkFailureSyncCondition(syncSetInstances)
+
+	// check if hivev1.SyncSetFailedCondition already set on the cluster deployment
+	currentCond := controllerutils.FindClusterDeploymentCondition(cd.Status.Conditions, hivev1.SyncSetFailedCondition)
+
+	if currentCond == nil && isfailedCondition == false {
+		return false, nil
+	}
+
+	// add/update hivev1.SyncSetFailedCondition to true
+	if isfailedCondition == true {
+		if currentCond == nil || currentCond.Status == corev1.ConditionFalse {
```

Comment:
the SetClusterDeploymentConditionWithChangeCheck() will do a lot of the work of only updating when necessary and notifying you if something changed (so you know whether to call the Status().Update() ). is there any particular reason to replicate some of the functionality in that call here instead of just calling SetClusterDeploymentConditionWithChangeCheck()?

Reply:
I am calling `SetClusterDeploymentCondition` which calls `SetClusterDeploymentConditionWithChangeCheck` to do the update. So finally the update does not happen is the condition is not getting changed  or the condition true. The code I have is complimentary to `SetClusterDeploymentConditionWithChangeCheck`. I have done similar in case of unreachable controller.

Reply:
The point is that there is no need to have all of this code here since `SetClusterDeploymentConditionWithChangeCheck` gives you everything that you need.

Everything after `cdLog.Infof("isFailedCondition: %v", isFailedCondition)` can be replaced with the following, which is more concise and arguably more readable
```
	status := corev1.ConditionFalse
	reason := "SyncSetApplySuccess"
	message := "SyncSet apply is successful"
	if isFailedCondition {
		status = corev1.ConditionTrue
		reason = "SyncSetApplyFailure"
		message = "One of the SyncSetInstance apply has failed"
	}
	conds, changed := controllerutils.SetClusterDeploymentConditionWithChangeCheck(
		cd.Status.Conditions,
		hivev1.SyncSetFailedCondition,
		status,
		reason,
		message,
		controllerutils.UpdateConditionNever,
	)
	if !changed {
		return false, nil
	}
	cd.Status.Conditions = conds
	if err := r.Status().Update(context.TODO(), cd); err != nil {
		cdLog.WithError(err).Error("error updating syncset failed condition")
		return false, err
	}
	return true, nil
```

Reply:
@staebler Agree. I am going to use the above.